### PR TITLE
Add SFS stress averaging to ABLStats

### PR DESCRIPTION
**Pull request by gantech**

### CODE REVIEW

File: amr-wind/wind_energy/ABLStats.cpp [hunk 1]
```diff
@@ -24,12 +25,16 @@ ABLStats::ABLStats(CFDSim& sim, const ABLWallFunction& abl_wall_func)
     , m_abl_wall_func(abl_wall_func)
     , m_temperature(sim.repo().get_field("temperature"))
     , m_mueff(sim.pde_manager().icns().fields().mueff)
+    , m_sfs(sim.repo().declare_field("sfs_stress", 3))
+    , m_t_sfs(sim.repo().declare_field("tsfs_stress", 3))
```
[the remark this answers is not in the document]
Comment (gantech) [reply amr-wind/wind_energy/ABLStats.cpp]: Yup..forgot about that. Will https://github.com/Exawind/amr-wind/pull/290/commits/c2d92951c3323ec334b1763e216e7c5b48beda27 work? Kinda looks ugly with `new` and `delete` lines. But basically tested it on N07 and seems to write the SFS stress out to the netcdf file. Thanks to you and Mike for designing the FPlaneAveraging class that takes ScratchFields!

File: amr-wind/wind_energy/ABLStats.H [hunk 2]
```diff
@@ -64,6 +64,9 @@ public:
         m_abl_forcing = forcing;
     }
 
+    //! Calculate sfs stress averages
+    void calc_sfs_stress_avgs();
+
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.H:69]: Why public?

File: amr-wind/wind_energy/ABLStats.H [hunk 3]
```diff
@@ -95,6 +98,8 @@ private:
     SecondMomentAveraging m_pa_tu;
     SecondMomentAveraging m_pa_uu;
     ThirdMomentAveraging m_pa_uuu;
+    ScratchFieldPlaneAveraging* m_pa_sfs;
+    ScratchFieldPlaneAveraging* m_pa_tsfs;
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.H:102]:
```suggestion
    std::unique_ptr<ScratchFieldPlaneAveraging> m_pa_sfs;
    std::unique_ptr<ScratchFieldPlaneAveraging> m_pa_tsfs;
```

Although I don't think these need to be instance variables.

File: amr-wind/wind_energy/ABLStats.cpp [hunk 4]
```diff
@@ -85,6 +86,7 @@ void ABLStats::calc_averages()
     m_pa_tu();
     m_pa_uu();
     m_pa_uuu();
+    calc_sfs_stress_avgs();
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.cpp:89]: These don't need to be calculated at every timestep... only if we have to output at a particular time instance correct? 

For that matter, the same applies to second and third moment averaging instances too.

File: amr-wind/wind_energy/ABLStats.cpp [hunk 5]
```diff
@@ -108,6 +110,9 @@ void ABLStats::post_advance_work()
     }
 
     process_output();
+
+    delete m_pa_sfs;
+    delete m_pa_tsfs;
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.cpp:115]:
```suggestion
```

File: amr-wind/wind_energy/ABLStats.cpp [hunk 6]
```diff
@@ -429,9 +440,85 @@ void ABLStats::write_netcdf()
...
+
+        {
+            amrex::Vector<std::string> var_names{
+                "u'theta'_sfs", "v'theta'_sfs", "w'theta'_sfs"};
+            for (int i = 0; i < AMREX_SPACEDIM; i++) {
+                m_pa_tsfs->line_average(i, l_vec);
+                auto var = grp.var(var_names[i]);
+                var.put(l_vec.data(), start, count);
+            }
+        }
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.cpp:462]: See comment below in `calc_sfs_stress_avgs`

File: amr-wind/wind_energy/ABLStats.cpp [hunk 7]
```diff
@@ -429,9 +440,85 @@ void ABLStats::write_netcdf()
...
     }
     ncf.close();
 #endif
 }
 
+void ABLStats::calc_sfs_stress_avgs()
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.cpp:468]: I would rename this `output_sfs_stress_avgs` and move the above section down here. That way you don't have to store instances of those averaging classes.

File: amr-wind/wind_energy/ABLStats.cpp [hunk 8]
```diff
@@ -429,9 +440,85 @@ void ABLStats::write_netcdf()
...
+        }
+    }
+
+    m_pa_sfs =
+        new ScratchFieldPlaneAveraging(*sfs_stress, m_sim.time(), m_normal_dir);
+    (*m_pa_sfs)();
+    m_pa_tsfs = new ScratchFieldPlaneAveraging(
+        *t_sfs_stress, m_sim.time(), m_normal_dir);
+    (*m_pa_tsfs)();
+}
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.cpp:522]: These can simply be 
```suggestion
    ScratchFieldPlaneAveraging pa_sfs(*sfs_stress, m_sim.time(), m_normal_dir);
    ScratchFieldPlaneAveraging pa_tsfs(*t_sfs_stress, m_sim.time(), m_normal_dir);
    pa_sfs();
    pa_tsfs();
    
    // Insert netcdf output here.
}
```

**Review by sayerhs — CHANGES_REQUESTED** [review]

File: amr-wind/wind_energy/ABLStats.cpp (hunk 8 above)
Comment (gantech) [reply amr-wind/wind_energy/ABLStats.cpp]: Tried addressing in https://github.com/Exawind/amr-wind/pull/290/commits/22cbdfe875f7c910a3fd1369c871a361c06ac0dd

File: amr-wind/wind_energy/ABLStats.cpp (hunk 7 above)
Comment (gantech) [reply amr-wind/wind_energy/ABLStats.cpp]: Tried addressing in https://github.com/Exawind/amr-wind/pull/290/commits/22cbdfe875f7c910a3fd1369c871a361c06ac0dd

File: amr-wind/wind_energy/ABLStats.cpp (hunk 6 above)
Comment (gantech) [reply amr-wind/wind_energy/ABLStats.cpp]: Tried addressing in https://github.com/Exawind/amr-wind/pull/290/commits/22cbdfe875f7c910a3fd1369c871a361c06ac0dd

File: amr-wind/wind_energy/ABLStats.cpp (hunk 5 above)
Comment (gantech) [reply amr-wind/wind_energy/ABLStats.cpp]: Tried addressing in https://github.com/Exawind/amr-wind/pull/290/commits/22cbdfe875f7c910a3fd1369c871a361c06ac0dd

File: amr-wind/wind_energy/ABLStats.cpp (hunk 4 above)
Comment (gantech) [reply amr-wind/wind_energy/ABLStats.cpp]: Tried addressing in https://github.com/Exawind/amr-wind/pull/290/commits/22cbdfe875f7c910a3fd1369c871a361c06ac0dd

File: amr-wind/wind_energy/ABLStats.H (hunk 3 above)
Comment (gantech) [reply amr-wind/wind_energy/ABLStats.H]: Tried addressing in https://github.com/Exawind/amr-wind/pull/290/commits/22cbdfe875f7c910a3fd1369c871a361c06ac0dd

File: amr-wind/wind_energy/ABLStats.H (hunk 2 above)
Comment (gantech) [reply amr-wind/wind_energy/ABLStats.H]: Tried addressing in https://github.com/Exawind/amr-wind/pull/290/commits/22cbdfe875f7c910a3fd1369c871a361c06ac0dd

File: amr-wind/wind_energy/ABLStats.H [hunk 9]
```diff
@@ -137,6 +133,11 @@ private:
     //! Number of cells in the horizontal direction
     size_t m_ncells_h1{0};
     size_t m_ncells_h2{0};
+
+#ifdef AMR_WIND_USE_NETCDF
+    //! Calculate and write sfs stress averages to output
+    void output_sfs_stress_avgs(ncutils::NCGroup grp, const size_t nt);
+#endif
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.H:140]: So looking at the logic in the code it seems like we can hide this from `ABLStats.H` and just make it a standalone function inside ABLStats.cpp

File: amr-wind/wind_energy/ABLStats.H [hunk 10]
```diff
@@ -4,6 +4,7 @@
 #include <memory>
 
 #include "amr-wind/CFDSim.H"
+#include "amr-wind/utilities/ncutils/nc_interface.H"
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.H:7]: If you eliminate `calc_sfs_stress_avgs` as a method you can eliminate this include

File: amr-wind/wind_energy/ABLStats.cpp [hunk 11]
```diff
@@ -441,34 +437,17 @@ void ABLStats::write_netcdf()
...
-        }
+        output_sfs_stress_avgs(grp, nt);
     }
     ncf.close();
 #endif
 }
 
-void ABLStats::calc_sfs_stress_avgs()
+#ifdef AMR_WIND_USE_NETCDF
+void ABLStats::output_sfs_stress_avgs(ncutils::NCGroup grp, const size_t nt)
```
Comment (sayerhs) [inline amr-wind/wind_energy/ABLStats.cpp:447]: We can make this a standalone function

Comment (gantech) [reply amr-wind/wind_energy/ABLStats.cpp]: Addressed in https://github.com/Exawind/amr-wind/pull/290/commits/da05df00d9686f43b32742e3ec13693811db1ab3

File: amr-wind/wind_energy/ABLStats.H (hunk 10 above)
Comment (gantech) [reply amr-wind/wind_energy/ABLStats.H]: Addressed in https://github.com/Exawind/amr-wind/pull/290/commits/da05df00d9686f43b32742e3ec13693811db1ab3

File: amr-wind/wind_energy/ABLStats.H (hunk 9 above)
Comment (gantech) [reply amr-wind/wind_energy/ABLStats.H]: Addressed in https://github.com/Exawind/amr-wind/pull/290/commits/da05df00d9686f43b32742e3ec13693811db1ab3

**Review by sayerhs — APPROVED** [review]

**Pull request merged**